scipy/fft/_fftlog_np.py

```diff
@@ -0,0 +1,186 @@
+import numpy as np
```

Comment:
@lucascolley I think this file is still to be removed, with the implementation moved back into `_fftlog.py` (a la the `_basic.py` / `_basic_uarray.py` split), right?

Reply:
I believe not. The difference here is that, for `_basic.py`, the backend which `uarray` uses is `pocketfft`. Since these functions were added more recently, and are not a part of `pocketfft`, their implementations are stored here in this file.

Reply:
To clarify, each function now has three versions: the public one, the multimethod and the NumPy backend.

For the basic functions, they are located in `_basic.py`, `_basic_uarray.py` and `pocketfft/...` respectively.
For those in fftlog, it is `_fftlog.py`, `_fftlog_multimethods.py` (which I already have down to rename for consistency), and `_fftlog_np.py`. Perhaps this file should be renamed as well, but I don't think that it can be removed.

Reply:
I had another look, and I believe my comment is correct. This code move isn't necessary, and it increases the diff by quite a lot. I think this needs doing still:
- rename `_fftlog_multimethods.py` to `_fftlog_uarray.py`
- move the contents of `_fftlog_np.py` back to their original place in `_fftlog.py`
- these are pure Python functions, so they can be implemented with the array API. The one exception is one usage of `special.poch`, but that may get its own array API support - and CuPy has a matching version. Did we discuss this before? I can't remember if we did. But it looks to me like this code should look like:

```python
def fht(...):
    xp = array_namespace(a)
    if is_numpy(xp):
        return _fftlog_uarray.fht(...)

    # Continue the initial implementation of `fht` here, replacing `np` by `xp
    ...
    return A
```

Reply:
I don't think that this will work.

> rename _fftlog_multimethods.py to _fftlog_uarray.py
move the contents of _fftlog_np.py back to their original place in _fftlog.py

The problem is that, then, the function which calls `_fftlog_uarray.fht` is the function which `_fftlog_uarray.fht` tries to dispatch to, giving an infinite loop where `is_numpy` is always satisfied and we keep going back to `_fftlog_uarray`.

This could probably be fixed by adding another parameter to the functions which skips straight to the implementation. Does that sound right, or is there a better way to do this?

Reply:
> these are pure Python functions, so they can be implemented with the array API.

I agree that this should be done though - working on it.

Reply:
`fftlog` is now array-agnostic https://github.com/scipy/scipy/pull/19005/commits/61eceffc2f2e207d7d626a5d7fd37511c079170b (bar PyTorch due to https://github.com/pytorch/pytorch/issues/59786).

Unfortunately, this has not helped to address the size of the diff due to my concerns above. I'll have a go at adding an extra parameter with the semantics of "already been through `uarray`".

Reply:
Adding a `from_uarray` parameter has worked :)) so the diff should have reduced in size quite a bit. I'll have a full look through now for any final bits then I think I've covered everything.

Reply:
Much nicer! I'll resolve this comment.